src/main/scala/loamstream/util/Sets.scala

```diff
@@ -0,0 +1,26 @@
+package loamstream.util
+
+/**
+ * @author clint
+ * Nov 9, 2018
+ */
+object Sets {
+  /**
+   * Compute the difference between two Sets efficiently.  Converts each parameter to a HashSet if they aren't already
+   * to take advantage of HashSet's faster diff() method.  This is needed because the previous approach, which used
+   * `--` had O(n^2) running time.  What's more, the default implementation of `diff` for non-HashSets just calls
+   * `--`, hence the need to ensure we have HashSets.  Empirically, using this method makes validation run in linear 
+   * time instead of quadratic time, as it did previously.   
+   */
+  def fasterSetDiff[A](lhs: Set[A], rhs: Set[A]): Set[A] = {
+    import scala.collection.immutable.HashSet
+    
+    def toHashSet(s: Set[A]): HashSet[A] = s match {
+      case hs: HashSet[A] => hs 
+      case _ => s.to[HashSet]
```

Comment:
Given that this is the implementation of `fasterSetDiff`, would it make sense to (over time) switch all the `Set` instances to `HashSet` and get rid of the conversion?

Reply:
Maybe?  I'm hesitant to do that without a bunch more profiling.  `Set.apply` gives you a variety of `Set` implementations depending on the set size, presumably for good reasons:
```
scala> Set(1).getClass
res2: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.Set$Set1

scala> Set(1,2).getClass
res3: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.Set$Set2

scala> Set(1,2,3).getClass
res5: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.Set$Set3

scala> (1 to 10).toSet.getClass
res6: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.HashSet$HashTrieSet
```
ie, it uses dumb non-general implementations for small set sizes, like
```scala
class Set2[A] private[collection] (elem1: A, elem2: A) ... {
  def contains(a: A): Boolean = a == elem1 || a == elem2
```
I'm guessing that's faster than some kind of hashtable or trie for small sizes, and it definitely uses less memory.  

TLDR: Switching everything to `HashSet` would add different overheads, and might have unintended consequences.

Reply:
I had no idea. How weird, but probably good for "typical" use cases.